Add directory page extension APIs

felixfbecker opened the pull request:
Extends our new view API to support registering views on directories. Adds a new ~~editor~~viewer called `DirectoryViewer`. The DirectoryViewer and Workspace are passed to the view provider.

PR that uses this API in Codecov: https://github.com/codecov/sourcegraph-codecov/pull/51

Implementation notes:
- DirectoryViewers are in the editor service, but don't have a model (just a URI), so no changes to the model service.
- I send the full context from the main thread instead of an ID to avoid race conditions with the editor not being synced yet. It would be nice to get this solved on an architectural level.
- It would also be nice to have comlink transfer handlers working.
- I named it `DirectoryViewer`, not `DirectoryEditor`, because a directory editor doesn't make sense. I didn't do a full rename of our internal editor service etc to a "viewer service" (so internally the `DirectoryViewer` is still an editor), because that can be easily changed later, while changing the public union type tag is a breaking.
- No support for context key expressions yet.
- Added some new type guards ✨

felixfbecker commented on shared/src/api/protocol/contribution.ts:
```diff
@@ -298,20 +298,26 @@ export interface SearchFilters {
 
 /** The containers to which an extension can contribute views. */
-export enum ContributableViewContainer {
+export const ContributableViewContainer = {
```
This avoids casts (enums have nominal typing)

lguychard commented on packages/sourcegraph-extension-api/src/sourcegraph.d.ts:
```diff
@@ -534,7 +534,7 @@ declare module 'sourcegraph' {
      * component it is ({@link CodeEditor}, etc.).
      */
-    export type ViewComponent = CodeEditor
+    export type ViewComponent = CodeEditor | DirectoryViewer
```
Historical moment - `ViewComponent` is no longer just an alias for `CodeEditor` 😄

lguychard commented on packages/sourcegraph-extension-api/src/sourcegraph.d.ts:
```diff
@@ -719,16 +743,53 @@ declare module 'sourcegraph' {
+
+    /**
+     * Experimental global view provider. Global view providers are shown on a dedicated page in the app.
+     * This API is experimtal subject to change or removal without notice.
```
```suggestion
     * This API is experimental and is subject to change or removal without notice.
```

felixfbecker replied:
🤦‍♂️

lguychard commented on shared/src/api/client/services/editorService.test.ts:
```diff
@@ -106,8 +106,7 @@ describe('EditorService', () => {
             })
-            const changes = editorService
-                .observeEditor(editorId)
+            const changes = (editorService.observeEditor(editorId) as Observable<CodeEditorData>)
```
Why is this cast now necessary?

felixfbecker replied:
`observeEditor()` returns `Observable<CodeEditor | DirectoryViewer>`, but we know that the editor ID above refers to a `CodeEditor`. I think this is a general problem of using IDs - it would be more type safe to use object references to refer to them.

lguychard commented on shared/src/api/client/services/viewService.ts:
```diff
@@ -1,47 +1,80 @@
 export interface View extends ExtensionView {}
 
+/**
+ * The types of the context parameter by the container (internally).
```
```suggestion
 * An internal map of container type to {@link ViewProvider} context parameter type.
```

lguychard commented on shared/src/api/extension/api/windows.ts:
```diff
@@ -96,13 +96,29 @@ export class ExtWindow implements sourcegraph.Window {
-                        this.documents
-                    )
+                    const { editorData } = update
+                    let editor: ExtCodeEditor | sourcegraph.DirectoryViewer
```
> I didn't do a full rename of our internal editor service etc to a "viewer service"

I think this will be sooner rather than later for clarity, I do a double take every time I see an `editor` that can also be a `DirectoryViewer`.

felixfbecker replied:
![It is done](https://media.giphy.com/media/3oKIPf3C7HqqYBVcCk/giphy.gif)

lguychard commented on shared/src/api/protocol/contribution.ts:
```diff
@@ -323,10 +329,7 @@ export interface ViewContribution {
-     *
-     * Currently only 'global' is supported; the other values of {@link ContributableViewContainer}
-     * are intended for the panel view API (which will eventually be merged into the new view API).
+     * Where this view will be displayed.
```
```suggestion
     * The container where this view will be displayed.
```

lguychard commented on shared/src/api/protocol/contribution.ts:
```diff
@@ -298,20 +298,26 @@ export interface SearchFilters {
+    GlobalPage: 'global/page',
+
+    /**
+     * A view on directory pages.
```
```suggestion
     * A view contributed to a directory page.
```

felixfbecker commented:
@lguychard could you take another look at the now-renamed version? :)
Renamed in https://github.com/sourcegraph/sourcegraph/pull/10236/commits/07a0a25dc5b41d106a7ff8554f381bfa537f9508

unknwon commented on schema/settings.schema.json:
```diff
@@ -18,6 +18,12 @@
           "!go": { "pointer": true }
         },
+        "codeInsights": {
+          "description": "Enables code insights on directory pages.",
```
nit: Maybe just me, I have no idea what "directory pages" means.

felixfbecker replied:
This is an example of a directory page: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/cmd/frontend

Do you have a suggestion for a better name?

unknwon replied:
> Do you have a suggestion for a better name?

Unfortunately, this seems the best name for me 😂 @christinaforney probably have a better name as a PM.

christinaforney replied:
I don't have a better name for you, "directory" is a common term in software to represent the container of other files. So, a "directory page" would be a page that shows the sub-files and directories and other data about the contents of the directory.

@unknwon - is that a good enough explanation for you to be comfortable with the naming?

unknwon replied:
@christinaforney thanks for the explanation! Do we have mention anywhere how the concept of "directory" fits into Sourcegraph? It is hard for me to link back which pages are considered as "directory pages" in Sourcegraph. Probably adding an example, something like "For example, https://sourcegraph.com/owner/repo is the directory page for the repository `owner/repo` in the repository root. Every subdirectory of the repository has their own directory pages."

I could imagine if I'm the site admin, I would have no idea what will happen after turning on this option.

Sorry about nitty-picky here 😛 I believe code insights will be extremely useful (not just what UI shows to users, but the APIs that one (i.e. me) could use to enable more interesting things on top of it).

felixfbecker replied:
@joe I think directory is the most accurate term. Here's why:
- As @christinaforney said, directories are generally an established term in software.
- Sourcegraph deals with Git repositories. Within a Git repository, subdirectories are _tree objects_. That is why on Sourcegraph and GitHub, subdirectories have `/tree/` in the URL.
- Both tree objects in Git and the repository itself map to _directories_ on the file system.
- Therefor the most accurate term to describe the pages that display _both_ the _repository_ (root) and _trees_ (subdirectories) is _directory page_, and `DirectoryViewer` in the extension API (the API is agnostic to whether its run in an IDE or on the web, so there is no concept of a "page", just a "viewer").

Adding the example is a good idea though. Note though that this is all nothing but a prototype for now, and once it gets out of that stage, I'll write actual documentation pages etc (and the feature flag will go away eventually too). I think for now it's not the best use of time to write extensive documentation of this. I don't expect any site admin to turn this on in its current form, there are way too many things that can break from the API changes.

unknwon replied:
Thanks, make senses!
